tests/contrib/__init__.py

```diff
@@ -41,6 +52,14 @@ def assert_not_double_wrapped(self, obj):
         self.assertTrue(hasattr(obj, '__wrapped__'), '{} is not wrapped'.format(obj))
         self.assert_not_wrapped(obj.__wrapped__)
 
+    def test_patched_library_not_imported(self):
+        """
+        TODO: this would be great to test, however we do not have a reliable
+        mechanism to "unimport" modules between each test case.
```

Comment:
reloading a module should trigger our hooks to run again.

```python
def import_module(self, name):
    if name in sys.modules:
        return reload_module(sys.modules[name])

     # or w/e this semantic is
    return importlib.import_module(name)

def tearDown(self):
    unpatch()
    self.assert_is_not_patched(module)

def test_case(self):
    patch(redis=True)
    redis = self.import_module('redis')

    self.assert_is_patched(redis)
```


This is just a very very rough idea.

After each test case we need to be make sure a module is back to it's original state.

The easiest way we can do this is disable any kind of patching hooks, and just reload the module.

We also need a way to trigger module hooks executing, which reloading a module should handle that just fine.


Sorry, these thoughts are kind of all over the place, but hoping this helps a little ?

Reply:
I like this approach. The only problem with not completely removing the module is that tests like `test_patch_before_import`, depending on test ordering, may already have the module loaded 😕

Reply:
Although I do see your point that reloading _should_ be the same as importing the module for the first time in terms of hooks and patching. Can we be certain that this is the case?

Reply:
If the module is loaded and we configure the hooks to patch.. it shouldn't patch until the next import/reload.